source/uwp/Renderer/lib/RenderedAdaptiveCard.cpp

```diff
@@ -432,42 +421,37 @@ namespace AdaptiveCards::Rendering::Uwp
         return m_inputs->AddInputValue(inputItem, renderArgs);
     }
 
-    HRESULT RenderedAdaptiveCard::LinkActionToCard(_In_ ABI::AdaptiveCards::Rendering::Uwp::IAdaptiveActionElement* action,
+    HRESULT RenderedAdaptiveCard::LinkActionToCard(_In_ ABI::AdaptiveCards::ObjectModel::Uwp::IAdaptiveActionElement* action,
                                                    _In_ ABI::AdaptiveCards::Rendering::Uwp::IAdaptiveRenderArgs* renderArgs)
     {
         return m_inputs->LinkSubmitActionToCard(action, renderArgs);
     }
 
-    InternalId GetInternalIdFromCard(_In_ ABI::AdaptiveCards::Rendering::Uwp::IAdaptiveCard* card)
-    {
-        ComPtr<AdaptiveCard> cardImpl = PeekInnards<AdaptiveCards::Rendering::Uwp::AdaptiveCard>(card);
-        return cardImpl->GetInternalId();
-    }
-
-    HRESULT RenderedAdaptiveCard::LinkCardToParent(_In_ ABI::AdaptiveCards::Rendering::Uwp::IAdaptiveCard* card,
+    HRESULT RenderedAdaptiveCard::LinkCardToParent(_In_ ABI::AdaptiveCards::ObjectModel::Uwp::IAdaptiveCard* card,
                                                    _In_ ABI::AdaptiveCards::Rendering::Uwp::IAdaptiveRenderArgs* renderArgs)
     {
         // We get the card internal id from the showcard action
-        InternalId cardId = GetInternalIdFromCard(card);
+        UINT32 cardId;
+        RETURN_IF_FAILED(card->get_InternalId(&cardId));
 
         // Then we get the parent card internal id from the renderArgs
         ComPtr<IAdaptiveRenderArgs> localRenderArgs(renderArgs);
-        InternalId parentCardId = InternalId();
+        UINT32 parentCardId = InternalId().Hash();
```

Comment:
I'm unsure how this type of using, but not storing, the InternalId may yield ids that get lost

Reply:
InternalId() with no args will return InternalId::Invalid, which represents the top level parent card. When you come in here with renderargs, we get the parent id from there (this is the case where you're linking to a show card). For the top level card renderargs will be null and we use the invalid id to indicate that the parent is the top level parent card.